File: ies/music-topos/src/music_topos/free_monad.clj
(ns music-topos.free-monad
  "Free Monad (Pattern) - Decision trees for musical actions.
   
   Based on Libkind & Spivak's 'Pattern Runs on Matter' (ACT 2024):
   - Free monad 𝔪p represents terminating decision trees
   - Pattern determines how a situation can unfold
   - Each position is a question, directions are possible answers
   
   Musical interpretation:
   - Pattern = the score/composition structure
   - Positions = musical choices (which note, which chord, which transformation)
   - Directions = specific instantiations of those choices"
  (:require [clojure.core.async :as async]))

;; ============================================================================
;; FREE MONAD DATA STRUCTURE
;; ============================================================================

;; Free f a = Pure a | Free (f (Free f a))
;; 
;; For music: f is the "instruction functor" - what musical operations exist
;; Pure = computation complete, return a value
;; Free = perform an operation, continue with the result

(defprotocol IFree
  "Protocol for free monad values"
  (pure? [this] "Is this a Pure value?")
  (run-free [this] "Extract the wrapped value or functor"))

(defrecord Pure [value]
  IFree
  (pure? [_] true)
  (run-free [this] (:value this)))

(defrecord Suspend [functor]
  IFree
  (pure? [_] false)
  (run-free [this] (:functor this)))

;; Smart constructors
(defn pure [x] (->Pure x))
(defn suspend [f] (->Suspend f))

;; ============================================================================
;; MUSICAL INSTRUCTION FUNCTOR
;; ============================================================================

;; The functor F for musical operations
;; Each constructor represents a musical "question" with continuations

(defrecord PlayNote [pitch duration amplitude next]
  ;; Play a single note, then continue
  ;; pitch: MIDI note number
  ;; duration: beats  
  ;; amplitude: 0-1
  ;; next: () -> Free F a (continuation after note plays)
  )

(defrecord PlayChord [pitches duration amplitude next]
  ;; Play multiple notes simultaneously
  ;; pitches: vector of MIDI notes
  ;; next: () -> Free F a
  )

(defrecord Rest [duration next]
  ;; Silent pause
  ;; next: () -> Free F a
  )

(defrecord Transform [transformation target next]
  ;; Apply a musical transformation (morphism in the topos)
  ;; transformation: keyword (:transpose, :invert, :retrograde, :plr-L, :plr-P, :plr-R)
  ;; target: the musical object to transform
  ;; next: (result) -> Free F a (continuation receives the transformed object)
  )

(defrecord Branch [condition then-branch else-branch]
  ;; Decision point based on musical context
  ;; condition: keyword for what to check (:in-tonic?, :cadence-ready?, etc.)
  ;; then-branch, else-branch: Free F a
  )

(defrecord Sequence [actions]
  ;; Sequential composition of multiple Free values
  ;; actions: vector of Free F a
  )

(defrecord Parallel [voices]
  ;; Parallel composition (polyphony)
  ;; voices: vector of Free F a to execute simultaneously
  )

;; ============================================================================
;; FUNCTOR INSTANCE (fmap)
;; ============================================================================

(defn fmap-instruction
  "Apply f to the continuation(s) of a musical instruction"
  [f instr]
  (condp instance? instr
    PlayNote   (->PlayNote (:pitch instr) (:duration instr) (:amplitude instr)
                           (comp f (:next instr)))
    PlayChord  (->PlayChord (:pitches instr) (:duration instr) (:amplitude instr)
                            (comp f (:next instr)))
    Rest       (->Rest (:duration instr) (comp f (:next instr)))
    Transform  (->Transform (:transformation instr) (:target instr)
                            (comp f (:next instr)))
    Branch     (->Branch (:condition instr) (f (:then-branch instr)) (f (:else-branch instr)))
    Sequence   (->Sequence (mapv f (:actions instr)))
    Parallel   (->Parallel (mapv f (:voices instr)))))

;; ============================================================================
;; MONAD INSTANCE
;; ============================================================================

(defn free-bind
  "Monadic bind for Free monad: Free f a -> (a -> Free f b) -> Free f b"
  [free-val f]
  (if (pure? free-val)
    (f (run-free free-val))
    (suspend (fmap-instruction #(free-bind % f) (run-free free-val)))))

(defn free-map
  "Functor map for Free monad"
  [f free-val]
  (free-bind free-val (comp pure f)))

(defn free-flatten
  "Join/flatten: Free f (Free f a) -> Free f a"
  [nested]
  (free-bind nested identity))

;; ============================================================================
;; DSL CONSTRUCTORS (lift instructions into Free)
;; ============================================================================

(defn play-note!
  "DSL: Play a note and continue"
  [pitch duration amplitude]
  (suspend (->PlayNote pitch duration amplitude (fn [] (pure nil)))))

(defn play-chord!
  "DSL: Play a chord and continue"
  [pitches duration amplitude]
  (suspend (->PlayChord pitches duration amplitude (fn [] (pure nil)))))

(defn rest!
  "DSL: Rest for a duration and continue"
  [duration]
  (suspend (->Rest duration (fn [] (pure nil)))))

(defn transform!
  "DSL: Apply transformation and receive result"
  [transformation target]
  (suspend (->Transform transformation target (fn [result] (pure result)))))

(defn branch!
  "DSL: Branch based on a condition"
  [condition then-branch else-branch]
  (suspend (->Branch condition then-branch else-branch)))

(defn sequence!
  "DSL: Sequential composition"
  [& actions]
  (suspend (->Sequence (vec actions))))

(defn parallel!
  "DSL: Parallel composition (polyphony)"
  [& voices]
  (suspend (->Parallel (vec voices))))

;; ============================================================================
;; EXAMPLE: INITIAL WORLD AS FREE MONAD
;; ============================================================================

(defn initial-world-pattern
  "The Initial World curriculum as a Free monad pattern (decision tree)"
  []
  (sequence!
   ;; 1. Initial pitch
   (play-note! 60 2.0 0.3)
   
   ;; 2. Chromatic scale (all 12 pitch classes)
   (sequence!
    (play-note! 60 0.08 0.2)
    (play-note! 61 0.08 0.2)
    (play-note! 62 0.08 0.2)
    (play-note! 63 0.08 0.2)
    (play-note! 64 0.08 0.2)
    (play-note! 65 0.08 0.2)
    (play-note! 66 0.08 0.2)
    (play-note! 67 0.08 0.2)
    (play-note! 68 0.08 0.2)
    (play-note! 69 0.08 0.2)
    (play-note! 70 0.08 0.2)
    (play-note! 71 0.08 0.2))
   
   ;; 3. Harmonic functions T → S → D → T
   (sequence!
    (play-chord! [60 64 67] 0.8 0.2)  ; Tonic (C major)
    (play-chord! [65 69 72] 0.8 0.2)  ; Subdominant (F major)
    (play-chord! [67 71 74] 0.8 0.2)  ; Dominant (G major)
    (play-chord! [60 64 67] 0.8 0.2)) ; Return to Tonic
   
   ;; 4. Circle of fifths modulation
   (sequence!
    (play-note! 60 0.5 0.25)  ; C
    (play-note! 67 0.5 0.25)  ; G (fifth up)
    (play-note! 62 0.5 0.25)  ; D (fifth up from G, mod 12)
    (play-note! 69 0.5 0.25)))) ; A

;; ============================================================================
;; ITERATEE / FOLD
;; ============================================================================

(defn fold-free
  "Fold a free monad given an interpreter for the functor.
   interpreter: F a -> M a (where M is some monad, e.g., IO or State)
   Returns the final value after running all instructions."
  [interpreter free-val]
  (if (pure? free-val)
    (run-free free-val)
    (interpreter (fmap-instruction #(fold-free interpreter %) (run-free free-val)))))

;; ============================================================================
;; PAIRING TYPE (for module structure with Cofree)
;; ============================================================================

(defprotocol IPairing
  "Pairing between functors f and g: (a -> b -> c) -> f a -> g b -> c
   This enables the module action: Free f ⊗ Cofree g → Free (f ⊗ g)"
  (pair [this combine f-val g-val]))

;; The pairing will be implemented when we create the Cofree comonad
;; It represents: for each question (f position), matter provides an answer (g direction)

File: ies/music-topos/lib/free_monad.rb
# lib/free_monad.rb
# Free Monad (Pattern) - Decision trees for musical actions
#
# Based on Libkind & Spivak's "Pattern Runs on Matter" (ACT 2024):
# - Free monad 𝔪p represents terminating decision trees
# - Pattern determines how a situation can unfold
# - Each position is a question, directions are possible answers

module FreeMonad
  # Base class for Free monad values
  class Free
    def pure?
      is_a?(Pure)
    end
    
    def bind(&f)
      raise NotImplementedError
    end
    
    def map(&f)
      bind { |a| Pure.new(f.call(a)) }
    end
    
    def flatten
      bind { |x| x }
    end
  end
  
  # Pure value - computation complete
  class Pure < Free
    attr_reader :value
    
    def initialize(value)
      @value = value
    end
    
    def bind(&f)
      f.call(@value)
    end
    
    def to_s
      "Pure(#{@value})"
    end
  end
  
  # Suspended computation - more work to do
  class Suspend < Free
    attr_reader :instruction
    
    def initialize(instruction)
      @instruction = instruction
    end
    
    def bind(&f)
      # fmap the bind into the instruction's continuation
      Suspend.new(@instruction.fmap { |free| free.bind(&f) })
    end
    
    def to_s
      "Suspend(#{@instruction.class.name.split('::').last})"
    end
  end
  
  # ============================================================================
  # MUSICAL INSTRUCTION FUNCTOR
  # ============================================================================
  
  # Base class for instructions
  class Instruction
    def fmap(&f)
      raise NotImplementedError
    end
  end
  
  # Play a single note
  class PlayNote < Instruction
    attr_reader :pitch, :duration, :amplitude, :next_fn
    
    def initialize(pitch, duration, amplitude, &next_fn)
      @pitch = pitch
      @duration = duration
      @amplitude = amplitude
      @next_fn = next_fn || -> { Pure.new(nil) }
    end
    
    def fmap(&f)
      PlayNote.new(@pitch, @duration, @amplitude) { f.call(@next_fn.call) }
    end
  end
  
  # Play a chord (multiple notes simultaneously)
  class PlayChord < Instruction
    attr_reader :pitches, :duration, :amplitude, :next_fn
    
    def initialize(pitches, duration, amplitude, &next_fn)
      @pitches = pitches
      @duration = duration
      @amplitude = amplitude
      @next_fn = next_fn || -> { Pure.new(nil) }
    end
    
    def fmap(&f)
      PlayChord.new(@pitches, @duration, @amplitude) { f.call(@next_fn.call) }
    end
  end
  
  # Rest (silence)
  class Rest < Instruction
    attr_reader :duration, :next_fn
    
    def initialize(duration, &next_fn)
      @duration = duration
      @next_fn = next_fn || -> { Pure.new(nil) }
    end
    
    def fmap(&f)
      Rest.new(@duration) { f.call(@next_fn.call) }
    end
  end
  
  # Apply a musical transformation
  class Transform < Instruction
    attr_reader :transformation, :target, :next_fn
    
    def initialize(transformation, target, &next_fn)
      @transformation = transformation
      @target = target
      @next_fn = next_fn || ->(result) { Pure.new(result) }
    end
    
    def fmap(&f)
      Transform.new(@transformation, @target) { |result| f.call(@next_fn.call(result)) }
    end
  end
  
  # Conditional branch
  class Branch < Instruction
    attr_reader :condition, :then_branch, :else_branch
    
    def initialize(condition, then_branch, else_branch)
      @condition = condition
      @then_branch = then_branch
      @else_branch = else_branch
    end
    
    def fmap(&f)
      Branch.new(@condition, f.call(@then_branch), f.call(@else_branch))
    end
  end
  
  # Sequential composition
  class Sequence < Instruction
    attr_reader :actions
    
    def initialize(actions)
      @actions = actions
    end
    
    def fmap(&f)
      Sequence.new(@actions.map { |a| f.call(a) })
    end
  end
  
  # Parallel composition (polyphony)
  class Parallel < Instruction
    attr_reader :voices
    
    def initialize(voices)
      @voices = voices
    end
    
    def fmap(&f)
      Parallel.new(@voices.map { |v| f.call(v) })
    end
  end
  
  # ============================================================================
  # DSL CONSTRUCTORS
  # ============================================================================
  
  module DSL
    def pure(x)
      Pure.new(x)
    end
    
    def play_note!(pitch, duration, amplitude)
      Suspend.new(PlayNote.new(pitch, duration, amplitude))
    end
    
    def play_chord!(pitches, duration, amplitude)
      Suspend.new(PlayChord.new(pitches, duration, amplitude))
    end
    
    def rest!(duration)
      Suspend.new(Rest.new(duration))
    end
    
    def transform!(transformation, target)
      Suspend.new(Transform.new(transformation, target))
    end
    
    def branch!(condition, then_branch, else_branch)
      Suspend.new(Branch.new(condition, then_branch, else_branch))
    end
    
    def sequence!(*actions)
      Suspend.new(Sequence.new(actions))
    end
    
    def parallel!(*voices)
      Suspend.new(Parallel.new(voices))
    end
  end
  
  # ============================================================================
  # FOLD / INTERPRETER
  # ============================================================================
  
  def self.fold(pattern, &interpreter)
    if pattern.pure?
      pattern.value
    else
      interpreter.call(pattern.instruction.fmap { |p| fold(p, &interpreter) })
    end
  end
end

# ============================================================================
# EXAMPLE: INITIAL WORLD AS FREE MONAD PATTERN
# ============================================================================

class InitialWorldPattern
  extend FreeMonad::DSL
  
  def self.build
    sequence!(
      # 1. Initial pitch
      play_note!(60, 2.0, 0.3),
      
      # 2. Chromatic scale
      sequence!(
        *((60..71).map { |p| play_note!(p, 0.08, 0.2) })
      ),
      
      # 3. Harmonic functions T → S → D → T
      sequence!(
        play_chord!([60, 64, 67], 0.8, 0.2),  # Tonic
        play_chord!([65, 69, 72], 0.8, 0.2),  # Subdominant
        play_chord!([67, 71, 74], 0.8, 0.2),  # Dominant
        play_chord!([60, 64, 67], 0.8, 0.2)   # Tonic return
      ),
      
      # 4. Circle of fifths
      sequence!(
        play_note!(60, 0.5, 0.25),  # C
        play_note!(67, 0.5, 0.25),  # G
        play_note!(62, 0.5, 0.25),  # D
        play_note!(69, 0.5, 0.25)   # A
      )
    )
  end
end
